File: src/memory/builder/resume_surface.py
"""Builder resume surface rendering."""

from __future__ import annotations

from memory.builder.lifecycle_ribbon import render_lifecycle_ribbon
from memory.builder.resume_state import BuilderResumeState
from memory.builder.roadmap_position import RoadmapPosition
from memory.builder.surface_protocol import wrap_ariad_surface


def render_builder_resume_surface(
    state: BuilderResumeState,
    *,
    roadmap_position: RoadmapPosition | None = None,
) -> str:
    """Render the Builder resume surface for an Ariad-governed journey."""
    cursor = state.cursor
    lines = [
        "Delivery",
        render_lifecycle_ribbon(_resume_phase(cursor.last_delivery_event if cursor else None)),
        "",
        "╭────────────────────────────────────────────────────────╮",
        "│        ■  BUILDER RESUME                               │",
        "│                                                        │",
        _card_text("journey"),
        _card_text(state.journey),
        "│                                                        │",
        _card_text("adopted method"),
        _card_text(state.adopted_method or "none"),
        "│                                                        │",
        _card_text("resumable"),
        _card_text("yes" if state.resumable else "no"),
    ]
    if state.reason:
        lines.extend(
            [
                "│                                                        │",
                _card_text("reason"),
                *_card_wrapped(state.reason),
            ]
        )
    lines.extend(
        [
            "│                                                        │",
            _card_text("roadmap position"),
            *_card_wrapped(_format_roadmap_position(roadmap_position)),
            "│                                                        │",
            _card_text("active item"),
            _card_text(cursor.active_item if cursor and cursor.active_item else "none"),
            "│                                                        │",
            _card_text("active checkpoint"),
            _card_text(cursor.active_checkpoint if cursor and cursor.active_checkpoint else "none"),
            "│                                                        │",
            _card_text("pending confirmation"),
            _card_text(
                cursor.pending_confirmation if cursor and cursor.pending_confirmation else "none"
            ),
            "│                                                        │",
            _card_text("last delivery event"),
            _card_text(
                cursor.last_delivery_event if cursor and cursor.last_delivery_event else "none"
            ),
            "│                                                        │",
            _card_text("allowed next actions"),
            *_card_prefixed(state.allowed_next_actions, "-"),
            "│                                                        │",
            _card_text("boundary"),
            *_card_wrapped("Builder resumes context only; no story lifecycle work was executed."),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("builder_resume", "\n".join(lines) + "\n")


def _resume_phase(last_delivery_event: str | None) -> str:
    if not last_delivery_event:
        return "pull"
    if "done" in last_delivery_event:
        return "done"
    if "coherence" in last_delivery_event:
        return "coherence"
    if "review" in last_delivery_event:
        return "review"
    if "validation" in last_delivery_event:
        return "validate"
    if "plan" in last_delivery_event:
        return "plan"
    if "prepare" in last_delivery_event or "template" in last_delivery_event:
        return "prepare"
    return "pull"


def _card_text(text: str) -> str:
    width = 54
    return f"│ {text[:width]:<{width}} │"


def _card_prefixed(items: tuple[str, ...], prefix: str) -> list[str]:
    if not items:
        return [_card_text("none")]
    lines: list[str] = []
    for item in items:
        wrapped = _wrap_plain_text(item, width=52)
        for index, line in enumerate(wrapped):
            marker = prefix if index == 0 else " "
            lines.append(_card_text(f"{marker} {line}"))
    return lines


def _card_wrapped(text: str) -> list[str]:
    return [_card_text(line) for line in _wrap_plain_text(text, width=54)]


def _wrap_plain_text(text: str, *, width: int) -> list[str]:
    words = text.split()
    lines: list[str] = []
    current = ""
    for word in words:
        if len(word) > width:
            if current:
                lines.append(current)
                current = ""
            for start in range(0, len(word), width):
                lines.append(word[start : start + width])
            continue
        candidate = f"{current} {word}".strip()
        if len(candidate) > width and current:
            lines.append(current)
            current = word
        else:
            current = candidate
    if current:
        lines.append(current)
    return lines or ["none"]


def _format_roadmap_position(position: RoadmapPosition | None) -> str:
    if position is None:
        return "none"
    return f"{position.code} — {position.title} ({position.status}) [{position.path}]"

File: ts/src/builder/resumeSurface.ts
// CV22.DS7.US8 plateau 1 — the `■ BUILDER RESUME` surface.
//
// Port of `src/memory/builder/resume_surface.py`, plus the pure half of
// `resume_state.py`: the three allowed-next-action tuples and the rule that picks
// between them. The DB-backed composition (`read_builder_resume_state`) arrives in
// plateau 2 with the delivery cursor, because the cursor's reader and writer share
// one serialization and D2's revert argument rests on those bytes being symmetric.
//
// Three conditional shapes a port merges by accident:
//
//   * the `reason` block renders only when a reason exists;
//   * `release intent` needs BOTH `release_intent` AND
//     `release_intent_delivery_story` — `and`, not `or`, so a cursor carrying one
//     of them shows neither;
//   * `_last_refinement_event` reports `"none"` whenever there is no active
//     Refinement Story, EVEN WHEN an event is recorded. Reading the field
//     directly is the obvious port and it is wrong.
//
// `_resume_phase` exists in Python and is never called by anything. It is not
// ported: dead code that the surface does not reach is not parity surface, and
// carrying it forward would import a maintenance obligation with no observable
// behavior. Recorded in the story's debt list instead.

import { cardPrefixed, cardText, cardWrapped } from "./card.ts";
import { CANONICAL_REFINEMENT_INDEX } from "./refinementField.ts";
import type { RoadmapScope } from "./roadmapScope.ts";
import { roadmapPositionLines } from "./scopePhrases.ts";
import { wrapAriadSurface } from "./surfaceProtocol.ts";

const FRAME_TOP = "╭────────────────────────────────────────────────────────╮";
const FRAME_BOTTOM = "╰────────────────────────────────────────────────────────╯";
const FRAME_BLANK = "│                                                        │";

/** Python `NO_ACTIVE_ITEM_ACTIONS`. */
export const NO_ACTIVE_ITEM_ACTIONS = [
  "inspect_roadmap",
  "pull_candidate_if_known",
  "inspect_method",
] as const;

/** Python `ACTIVE_ITEM_ACTIONS`. */
export const ACTIVE_ITEM_ACTIONS = [
  "prepare_active_item",
  "inspect_roadmap",
  "inspect_method",
] as const;

/** Python `PENDING_CONFIRMATION_ACTIONS`. */
export const PENDING_CONFIRMATION_ACTIONS = [
  "answer_pending_confirmation",
  "inspect_method",
] as const;

/** The cursor fields the resume surface reads. */
export interface ResumeCursorView {
  readonly activeItem: string | null;
  readonly activeCheckpoint: string | null;
  readonly pendingConfirmation: string | null;
  readonly lastDeliveryEvent: string | null;
  readonly releaseIntent: string | null;
  readonly releaseIntentDeliveryStory: string | null;
}

/** Python `BuilderResumeState`. */
export interface BuilderResumeState {
  readonly journey: string;
  readonly adoptedMethod: string | null;
  readonly cursor: ResumeCursorView | null;
  readonly resumable: boolean;
  readonly reason: string | null;
  readonly allowedNextActions: readonly string[];
}

/**
 * Python's branch inside `read_builder_resume_state`: pending confirmation wins
 * over an active item, which wins over neither.
 */
export function selectAllowedNextActions(cursor: ResumeCursorView): readonly string[] {
  if (cursor.pendingConfirmation) return PENDING_CONFIRMATION_ACTIONS;
  if (cursor.activeItem) return ACTIVE_ITEM_ACTIONS;
  return NO_ACTIVE_ITEM_ACTIONS;
}

/** Python `_release_intent_lines`: both fields, or no block at all. */
function releaseIntentLines(cursor: ResumeCursorView | null): string[] {
  const intent = cursor?.releaseIntent ?? null;
  const deliveryStory = cursor?.releaseIntentDeliveryStory ?? null;
  if (!intent || !deliveryStory) return [];
  return [
    FRAME_BLANK,
    cardText("release intent"),
    ...cardWrapped(`${deliveryStory}: ${intent}`),
    ...cardWrapped("intent is not release authorization"),
  ];
}

/** Python `_refinement_field_lines`. Two shapes since CV22.DS10.TS4. */
function refinementFieldLines(canonicalRefinementIndex: string | null): string[] {
  if (canonicalRefinementIndex) {
    return [
      cardText("authority: project files"),
      ...cardWrapped(`index: ${canonicalRefinementIndex}`),
    ];
  }
  return [
    cardText("authority: project files (not started)"),
    ...cardWrapped(`create: ${CANONICAL_REFINEMENT_INDEX}`),
  ];
}

/**
 * Python `render_builder_resume_surface`, except for the `roadmap position` row.
 *
 * Python filled that row with the first roadmap file whose status contained
 * "Active", whichever journey asked. Since CR002 it states the journey's own
 * scope, which the caller resolves from the same cursor `state` carries. The
 * scope is required so that no caller can fall back to a guess by omission.
 */
export function renderBuilderResumeSurface(
  state: BuilderResumeState,
  options: {
    scope: RoadmapScope;
    canonicalRefinementIndex?: string | null;
  },
): string {
  const canonicalRefinementIndex = options.canonicalRefinementIndex ?? null;
  const cursor = state.cursor;
  const lines: string[] = [
    "Delivery",
    "",
    FRAME_TOP,
    "│        ■  BUILDER RESUME                               │",
    FRAME_BLANK,
    cardText("journey"),
    cardText(state.journey),
    FRAME_BLANK,
    cardText("adopted method"),
    cardText(state.adoptedMethod || "none"),
    FRAME_BLANK,
    cardText("resumable"),
    cardText(state.resumable ? "yes" : "no"),
  ];
  if (state.reason) {
    lines.push(FRAME_BLANK, cardText("reason"), ...cardWrapped(state.reason));
  }
  lines.push(
    FRAME_BLANK,
    cardText("roadmap position"),
    ...roadmapPositionLines(options.scope).flatMap(cardWrapped),
    FRAME_BLANK,
    cardText("active item"),
    cardText(cursor?.activeItem || "none"),
    FRAME_BLANK,
    cardText("active checkpoint"),
    cardText(cursor?.activeCheckpoint || "none"),
    FRAME_BLANK,
    cardText("pending confirmation"),
    cardText(cursor?.pendingConfirmation || "none"),
    FRAME_BLANK,
    cardText("last delivery event"),
    cardText(cursor?.lastDeliveryEvent || "none"),
    ...releaseIntentLines(cursor),
    FRAME_BLANK,
    cardText("🧰 Refinement field"),
    ...refinementFieldLines(canonicalRefinementIndex),
    FRAME_BLANK,
    cardText("allowed next actions"),
    ...cardPrefixed(state.allowedNextActions, "-"),
    FRAME_BLANK,
    cardText("boundary"),
    ...cardWrapped("Builder resumes context only; no story lifecycle work was executed."),
    FRAME_BOTTOM,
  );
  return wrapAriadSurface("builder_resume", `${lines.join("\n")}\n`);
}
